src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -0,0 +1,646 @@
+/*-------------------------------------------------------------------------
+ *
+ * shared_connection_stats.c
+ *   Keeps track of the number of connections to remote nodes across
+ *   backends. The primary goal is to prevent excessive number of
+ *   connections (typically > max_connections) to any worker node.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "pgstat.h"
+
+#include "libpq-fe.h"
+
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "access/htup_details.h"
+#include "catalog/pg_authid.h"
+#include "commands/dbcommands.h"
+#include "distributed/cancel_utils.h"
+#include "distributed/connection_management.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/shared_connection_stats.h"
+#include "distributed/time_constants.h"
+#include "distributed/tuplestore.h"
+#include "utils/builtins.h"
+#include "utils/hashutils.h"
+#include "utils/hsearch.h"
+#include "storage/ipc.h"
+
+
+#define REMOTE_CONNECTION_STATS_COLUMNS 4
+
+#define ADJUST_POOLSIZE_AUTOMATICALLY 0
+#define DISABLE_CONNECTION_THROTTLING -1
+
+/*
+ * The data structure used to store data in shared memory. This data structure is only
+ * used for storing the lock. The actual statistics about the connections are stored
+ * in the hashmap, which is allocated separately, as Postgres provides different APIs
+ * for allocating hashmaps in the shared memory.
+ */
+typedef struct ConnectionStatsSharedData
+{
+	int sharedConnectionHashTrancheId;
+	char *sharedConnectionHashTrancheName;
+
+	LWLock sharedConnectionHashLock;
+	ConditionVariable waitersConditionVariable;
+} ConnectionStatsSharedData;
+
+typedef struct SharedConnStatsHashKey
+{
+	/*
+	 * In some cases, Citus allows opening connections to hosts where
+	 * there is no notion of "WorkerNode", such as task-tracker daemon.
+	 * That's why, we prefer to use "hostname/port" over nodeId.
+	 */
+	char hostname[MAX_NODE_LENGTH];
+	int32 port;
+
+	/*
+	 * Given that citus.shared_max_pool_size can be defined per database, we
+	 * should keep track of shared connections per database.
+	 */
+	Oid databaseOid;
```

Comment:
@marcocitus you told me that you want to think a bit more about `databaseOid` being here. I still think that it should be in it, as `citus.shared_max_connections` is per database. 

Still reminding to you in case you forgot about this and I'm missing something in my assesment.